base/compiler/typeinfer.jl

```diff
@@ -227,7 +227,11 @@ function finish!(interp::AbstractInterpreter, caller::InferenceResult)
 end
 
 function _typeinf(interp::AbstractInterpreter, frame::InferenceState)
-    typeinf_nocycle(interp, frame) || return false # frame is now part of a higher cycle
+    ccall(:jl_typeinf_begin, Cvoid, ())
```

Comment:
This is still a very wide lock. You should only require it inside `cache_result!` right now

Reply:
If the typeinf lock is now scoped to just `cache_result!` rather than all of type inference, is that sufficient to deprioritize the typeinf lock to be below the codegen lock?